lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,249 @@
+func (og OpsgenieClient) CreateAlert(ctx context.Context, reqID string, reqData RequestData) (OpsgenieData, error) {
+	bodyDetails, err := buildAlertBody(og.WebProxyURL, reqID, reqData)
+	if err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	body := AlertBody{
+		Message:     fmt.Sprintf("Access request from %s", reqData.User),
+		Alias:       fmt.Sprintf("%s/%s", ogAlertKeyPrefix, reqID),
+		Description: bodyDetails,
+		Responders:  og.getResponders(reqData),
+		Priority:    og.Priority,
+	}
+
+	var result AlertResult
+	if _, err = og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetResult(&result).
+		Post("alerts"); err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	return OpsgenieData{
+		AlertID: result.Alert.ID,
+	}, nil
+}
+
+func (og OpsgenieClient) getResponders(reqData RequestData) []Responder {
+	schedules := og.DefaultSchedules
+	if reqSchedules, ok := reqData.RequestAnnotations[ReqAnnotationRespondersKey]; ok {
+		schedules = reqSchedules
+	}
+	responders := []Responder{}
+	for _, s := range schedules {
+		responders = append(responders, Responder{
+			Type: "schedule",
+			ID:   s,
+		})
+	}
+	return responders
+}
+
+// PostReviewNote posts a note once a new request review appears.
+func (og OpsgenieClient) PostReviewNote(ctx context.Context, alertID string, review types.AccessReview) error {
+	note, err := buildReviewNoteBody(review)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/notes"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ResolveAlert resolves an alert and posts a note with resolution details.
+func (og OpsgenieClient) ResolveAlert(ctx context.Context, alertID string, resolution Resolution) error {
+	note, err := buildResolutionNoteBody(resolution)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/close"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// GetOnCall returns the list of responders on-call for a schedule.
+func (og OpsgenieClient) GetOnCall(ctx context.Context, scheduleName string) (RespondersResult, error) {
+	// v2/schedules/ScheduleName/on-calls?scheduleIdentifierType=name&flat=true'
+
+	var result RespondersResult
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetPathParams(map[string]string{"scheduleName": scheduleName}).
+		SetQueryParams(map[string]string{
+			"scheduleIdentifierType": "name",
+			// When flat is enabled it returns user names of on-call participants.
```

Comment:
I think it returns emails and not user names
https://docs.opsgenie.com/docs/who-is-on-call-api#get-on-calls